research/automated_release.md

```diff
@@ -0,0 +1,148 @@
+
+It proscribes two long-lived branches, with other, shorter, branches forking from one, and merging back into one, or both.
+
+The `develop` branch is the core branch for development. A new `feature` branch is forked from here for each task. When the task is done, the branch is merged back into develop.
+
+In theory, this branch interaction model is sufficient for multiple programmers to cooperatively develop software. In practice, develop is not sufficiently stable for release to end users.
+
+`master` is the branch intended for stable software releases. This branch should only receive code that is thoroughly tested, and does not contain partially finished features.
+
+In order to verify that code on `develop` is ready for primetime, new `release` branches are created regularly. These fork from `develop` either after a set interval (end of Scrum sprint), or when a feature is complete.
+
+At this point developers can continue to add new features to `develop`, while the `release` branch is used to verify that its frozen set of features are bug-free, and meet requirements.
+
+Bug fixes made on `release` branches should also be merged into develop. No new functionality should be added to `release`.
+
+When the software on `release` passes all checks, it is merged into `master`.
+
+Occasionally, despite best efforts, `master` will have bugs. For the most critical of them, `hotfix` branches are forked directly from `master` to fix the issue.
+When ready, these branches are forked both into `master`, and `develop`.
+
+Brewblox as a project has no special requirements that mandate any changes to this model. At time of writing, we do still need to implement meaningful and automatic regression tests that should be run against `release` software.
+
+We do need to formulate a standard approach for conditions not described explicitly by Git Flow:
+
+* Contributors working on forked repositories
+* Pull Requests
+
+## Git Flow: Forked
+
+The default Git Flow model does not describe forked repositories making upstream pull requests. The forked model is attractive to projects with open-source contributors, as it allows contributions without granting write access to strangers.
+
+Contributors fork the base repository on GitHub into their own. At that point they can create their own branches, or use their local version of `develop` - the upstream repository doesn't care. When they are done, they make a pull request to the correct target branch on the base repository. Pull requests can be commented on by many, but only merged by somebody with write access to the base repository.
+
+We'll refer to the base repository as `upstream`, and to the contributor repository as `downstream`.
+
+In a forked Git Flow repository, there will be no `feature` or `hotfix` branches in `upstream`. This is all done in `downstream`.
+
+Contributors fork new `downstream/feature` branches from `upstream/develop`, and when ready, make a pull request back into `upstream/develop`.
```

Comment:
forking -> branching

Reply:
fixed everywhere: branch creation is now referred to as branching.